### ISSUE TITLE
Installation of the ruby agent (APM) is underdocumented

### ISSUE THREAD

**Issue by x-yuri** [newrelic/newrelic-ruby-agent#1713]
## Description
First, apparently `newrelic_rpm` should be required after all other gems. Otherwise it might not discover a dispatcher, or not install some instrumentation. What I saw people are doing is adding `NewRelic::Agent.manual_start`, in addition to implicitly requiring `newrelic_rpm`. I'm not sure if it adds something, or have some negative side effects, but it sounds like a kludge to me. I think it's best to mark the gem with `require: false` and require it manually after `Bundler.require :default` or whathaveyou.

Second, it doesn't always work with `puma` + `sinatra` websites. I see the following cases here:

* single mode (no workers). `newrelic_rpm` discovers the dispatcher (`puma`). `puma` is considered a forking dispatcher, as such the agent doesn't start.

  To make it work one needs to add `NewRelic::Agent.after_fork` after requiring `newrelic_rpm`. `force_reconnect` doesn't seem to be needed, since there's just one process and `newrelic_rpm` doesn't start the agent automatically.

* cluster mode (master + workers). `puma` is not discovered (because `puma` changes $0 of the workers to something along the lines of `puma: cluster worker n: ...`), as such the agent starts in each worker.

* cluster mode + `preload_app` (the app is loaded in the master process, not in the workers). `puma` is discovered, as such the agent doesn't start.

  To make it work one needs to add the following code to the puma config:

  ```ruby
  on_worker_boot do
    NewRelic::Agent.after_fork
  end
  ```

  `force_reconnect` doesn't seem to be needed, since the agent is not started in the master process.

Correct me if I'm wrong.

So it's not just a matter of adding `newrelic_rpm` to the `Gemfile`. And I bet this to some extent concerns Ruby on Rails websites as well.

## Expected Behavior
I'd expect the installation process to be better documented.

## For Maintainers Only or Hero Triaging this bug
*Suggested Priority (P1,P2,P3,P4,P5):*
*Suggested T-Shirt size (S, M, L, XL, Unknown):*

**Comment by x-yuri** [comment]
I've experimented a bit more. Let's take this simple `sinatra` app:

`docker-compose.yml`:

```yaml
services:
  app:
    build: .
    command: sleep infinity
    init: yes
    ports:
      - 9292:9292
    volumes:
      - .:/app
      - bundle:/usr/local/bundle
volumes:
  bundle:
```

`Dockerfile`:

```dockerfile
FROM ruby:2.6-alpine3.15
WORKDIR /app
```

`Gemfile`:

```rb
source 'https://rubygems.org'
gem 'newrelic_rpm', '8.14.0', require: false
gem 'sinatra', '3.0.5'
gem 'puma', '6.0.0'
```

`config.ru`:

```rb
require './app'
run Sinatra::Application
```

`app.rb`:

```rb
Bundler.require :default
require 'newrelic_rpm'

get '/' do
  'Hello world!'
end
```

`config/puma.rb`:

```rb
if ENV['PUMA_PRELOAD_APP'].to_i > 0
  preload_app!
end
if ENV['PUMA_WORKERS'].to_i > 0
  workers ENV['PUMA_WORKERS'].to_i
end
```

and consider ways to run it:

* `bundle exec puma` (single mode), `puma` discovered -> [starts][a] on the [first request][b]
* `PUMA_WORKERS=2 bundle exec puma` (cluster mode, `preload_app`: -), `puma` not discovered -> starts right away in the workers
* `PUMA_WORKERS=2 PUMA_PRELOAD_APP=1 bundle exec puma` (cluster mode, `preload_app`: +), `puma` discovered -> starts on the first request in the workers
* `bundle exec rackup -o 0.0.0.0` (single mode), `puma` not discovered -> starts right away
* `PUMA_WORKERS=2 bundle exec rackup -o 0.0.0.0` (cluster mode), `puma` not discovered -> starts right away in the master, but on the first request starts in the workers as well (with `rackup` it's like a forced `preload_app` mode, no point in telling `puma` to preload the app)

Single mode - one process, cluster mode - one master process and many worker processes, `preload_app` - preload the app (`config.ru`) in the master process.

Considering this it seems like `NewRelic::Agent.after_fork` doesn't need to be called manually (from the user code), as opposed to what I initially thought. Correct me if I'm wrong.

What's left is the load `newrelic` last thing. Apparently when `newrelic_rpm` is required it instruments what's already loaded. If you later call `NewRelic::Agent.manual_start`, it [won't instrument][c] the rest. Discovering a dispatcher doesn't seem like an important thing. Which brings up a question, is there a point in calling `NewRelic::Agent.manual_start`? At least in this case, and maybe generally. Is it safe to do so after requireing `newrelic_rpm`?

For now it looks like requiring `newrelic_rpm` last (after `ruby` libraries and gems) is the best possible approach. That should be enough to make the New Relic Ruby agent work.

[a]: https://github.com/newrelic/newrelic-ruby-agent/blob/8.14.0/lib/new_relic/agent/harvester.rb#L22-L28
[b]: https://github.com/newrelic/newrelic-ruby-agent/blob/8.14.0/lib/new_relic/agent/harvester.rb#L47
[c]: https://github.com/newrelic/newrelic-ruby-agent/blob/8.14.0/lib/new_relic/control/instrumentation.rb#L53-L55

**Comment by x-yuri** [comment]
Upon further investigation it appears that `newrelic_rpm` can be the first gem, and it still will instrument things. Because `rack` will be most likely loaded at the point where `newrelic_rpm` is required. That makes `newrelic_rpm` being able to schedule [another round of instrumentation][a] for the moment where all libraries are likely to be loaded. Which means that simply adding `newrelic_rpm` to the `Gemfile` should do.

Still it remains unclear what `NewRelic::Agent.manual_start` can do and when it should be used. At the moment it looks like a "if it doesn't work, try rebooting you machine" kind of solution.

[a]: https://github.com/newrelic/newrelic-ruby-agent/blob/8.14.0/lib/new_relic/agent/instrumentation/rack/instrumentation.rb#L36-L45

**Comment by tannalynn** [comment]
@x-yuri Thanks for bringing attention to areas of our docs that need improvement! We will take a look at how to make it more clear how the agent works in general and also when using sinatra. I'll update the label to documentation so we can take a look that. 

Like you pointed out, the deferred dependency detection makes it so the placement of `newrelic_rpm` in the gemfile isn't as much of an issue. However, there are still issues that pop up every once in a while that can cause odd or incorrect behavior unless the agent is listed last in the gemfile. I believe dispatcher issues (like you discussed in a previous comment) are one of the most common situations that can be fixed by moving the agent to the bottom of the gemfile. The dispatcher being accurate is important in situations where a forking dispatcher (like puma or unicorn) is being used. If the agent detects a forking dispatcher, it will defer connecting to the new relic servers until after forking (using the after_fork method you mentioned previously). 

As for when `NewRelic::Agent.manual_start` should be used, it is used in situations where the agent will not automatically start. This is typically due to agent configuration (disabling the autostart of the agent to later manually start it) or due to not using a supported framework that will kick off the auto start process, such as a pure ruby script. Although rare, occasionally an application may be doing something unexpected with a framework that can cause the agent autostart hooks to fail to trigger, which would make a manual start necessary in this situation as well.

Please let me know if there is anything else I can help with!

**Comment by x-yuri** [comment]
Let me sum it up. I saw in a project's code `NewRelic::Agent.manual_start`, and thought, "Is it really needed?" So I tried to look into how it works. And discovered that the dispatcher is not always detected. But then I saw that the agent starts anyway on the first request. Also I saw that instrumentation is made to not instrument things twice, but then I found that it patches `rack` to instrument what's left before starting the server. And at this point the agent looks smart enough to start w/o `NewRelic::Agent.manual_start` in the cases I considered (single process or master + workers, `puma` or `rackup` + `puma`) regardless of when it's required. But you say that discovering a dispatcher is important, so maybe I'm missing something?

If discovering a dispatcher is really important, then [your check][a] is pretty naive. Sometimes `$0` is `puma` (`bundle exec puma`, cluster mode + master process or single mode), sometimes `puma: cluster worker n: ...` (`bundle exec puma`, cluster mode + worker process), sometimes `rackup` (`bundle exec rackup`), sometimes `app.rb` or something (`bundle exec ruby app.rb`). And `puma` might not be required yet. It's not like there's a sure way to make certain that the dispatcher is `puma`, and my guess here is that the agent is made to work correctly even if the dispatcher is not discovered, and that dispatcher is `puma` running in cluster mode.

Returning to the question if the documentation is good enough, the main concern I have at the moment is that it says:

> Assuming you have installed the agent per above, you may also need to tell the Ruby agent to start for some frameworks and non-framework environments.

Now, how do I know if I need `NewRelic::Agent.manual_start`? One way is to probably look in the log for:

> ... Reporting to: https://rpm.newrelic.com/...

Another is to wait for the app to appear on the APM page. But if the agent is installed and I remove `NewRelic::Agent.manual_start`, how do I know if the data keep coming?

On a side note, I'm not sure if the agent is needed in development and test environments. And I wonder about the way to disable it. The obvious one is to just not require it. The better option is probably to add `agent_enabled: no` to `config/newrelic.yml`. But there are other `*enabled` options in the config. So I wonder if `agent_enabled: no` disables it all. Apparently yes, but just to be sure.

[a]: https://github.com/newrelic/newrelic-ruby-agent/blob/8.14.0/lib/new_relic/local_environment.rb#L146-L148
